File: pp/Final/ExplicitClosure.java

interface Closure<A,B> {
    public B apply(A arg);
};

class Multiply implements Closure<Integer,Integer> {
    Integer n;
    Multiply(Integer n0) {
        n = n0;
    };
    public Integer apply(Integer arg) {
        return n*arg;
    };
};


class Add implements Closure<Integer,Integer> {
    Integer n;
    Add(Integer n0) {
        n = n0;
    };
    public Integer apply(Integer arg) {
        return n+arg;
    };
};

interface List<A> {
    public void print();
    public <B> List<B> map(Closure<A,B> f);
};

class Nil<A> implements List<A> {
    public <B> List<B> map(Closure<A,B> f) {
        return new Nil<B>();
    }
    
    public void print() {
        System.out.println("empty.");
    }

}

class Cons<A> implements List<A> {
    A head;
    List<A> tail;
    public Cons (A h, List<A> t) {
        head = h;
        tail = t;
    }
    public <B> List<B> map(Closure<A,B> f) {
        return new Cons<B>(f.apply(head),tail.map(f));
        
    }
    public void print() {
        System.out.println("Element: " + head.toString());
        tail.print();
    }
}

public class ExplicitClosure {
    public static void main(String args[]) {
        List<Integer> l = new Cons<Integer>(1234, new Cons<Integer>(5678, new Nil<Integer>()));
        // l.print();
        l.map(new Add(2)).print();
    }
}

File: pp/Final/ExplicitClosure.cpp
#include <cstdio>

struct closure {
  int (* p)(int arg, int* env);
  int* env;
};

int apply(closure* c, int arg) {
  return c->p(arg,c->env);
}

void map(closure* f, int sz, int* a) {
  for (int i = 0; i<sz; i++) {
    a[i] = apply(f,a[i]);
  }
}

int add3(int arg, int* env) {
  return arg + 3;
}

int addN(int arg, int* n) {
  return arg + *n;
}

int main() {
  closure test1 = {add3};

  // int k = 15;
  // closure test2 = {addN, &k};

  int arr[5]  = {2,3,4,5,6};
  map(&test1,5,arr);
  for (int i = 0; i < 5; i++) {
    printf("%d\n",arr[i]);
  }
  return 0;
}



